sdks/python/apache_beam/runners/interactive/caching/streambasedcache.py

```diff
@@ -0,0 +1,491 @@
+        ]
+      self._finalizers += [
+          finalize(self, delete_subscriptions, self._child_subscriptions),
+          finalize(self, pub_client.delete_topic, self.topic.name),
+      ]
+
+  def _assert_topic_exists(self):
+    pub_client = pubsub.PublisherClient()
+    try:
+      _ = pub_client.get_topic(self.topic.name)
+    except gexc.NotFound:
+      raise IOError("Pubsub topic '{}' does not exist.".format(self.topic.name))
+
+  def _create_child_subscription(self, seek_to_start=True):
+    sub_client = pubsub.SubscriberClient()
+    sub_path = None
+    existing_sub_paths = [
+        sub.name for sub in ([self.subscription] + self._child_subscriptions)
+    ]
+    while sub_path is None or sub_path in existing_sub_paths:
+      sub_path = sub_client.subscription_path(
+          self.project, self.topic_name + '-' + uuid.uuid4().hex)
+    if not seek_to_start or self.snapshot is not None:
+      subscription = sub_client.create_subscription(sub_path, self.location)
+      self._child_subscriptions.append(subscription)
+      if seek_to_start:
+        sub_client.seek(
+            subscription.name,
+            snapshot=self.snapshot.name,
+            retry=api_core.retry.Retry())
+    elif not self._primary_sub_exhausted:
+      subscription = self.subscription
+      self._primary_sub_exhausted = True
+    else:
+      raise ValueError(
+          "Cannot seek_to_start more than once in an environment where "
+          "snapshots are not supported.")
+    return subscription
+
+
+class PatchedPubSubReader(PTransform):
+
+  def __init__(self, cache, reader_class, *reader_args, **reader_kwargs):
+    self._cache = cache
+    self._reader_class = reader_class
+    self._reader_args = reader_args
+    self._reader_kwargs = reader_kwargs
+
+    if "timestamp_attribute" not in reader_kwargs:
+      raise ValueError(
+          "timestamp_attribute must be specified when reading from cache.")
+
+  def expand(self, pbegin):
+    reader_kwargs = self._reader_kwargs.copy()
+
+    reader = self._reader_class(*self._reader_args, **reader_kwargs)
+    if reader_kwargs.get("with_attributes"):
+      return pbegin | reader
+    else:
+      return pbegin | reader | beam.Map(self._cache.coder.decode)
+
+
+class PatchedPubSubWriter(PTransform):
+
+  def __init__(self, cache, writer_class, writer_args, writer_kwargs):
+    self._cache = cache
+    self._writer_class = writer_class
+    self._writer_args = writer_args
+    self._writer_kwargs = writer_kwargs
+
+  def expand(self, pcoll):
+
+    if self._cache.requires_coder and self._cache.coder is None:
+      self._cache.coder = coders.registry.get_coder(pcoll.element_type)
+
+    coder = self._cache.coder
+    writer_kwargs = self._writer_kwargs.copy()
+    attributes_fn = writer_kwargs.pop("attributes_fn", None)
+    # DirectRunner does not support timestamp_attribute
+    timestamp_attribute = writer_kwargs.pop("timestamp_attribute", "ts")
```

Comment:
Please use the _default_timestamp_attribute constant.

Reply:
Done.